File: encryption-component/src/main/java/infrastructure/EncryptUtils.java
package infrastructure;

import javax.crypto.Cipher;
import java.security.KeyFactory;
import java.security.MessageDigest;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Arrays;
import java.util.Base64;

public class EncryptUtils {

    public static String encrypt(String data, String publicKey) {
        Cipher rsa;
        byte[] encryptedByte = new byte[0];
        try {
            rsa = Cipher.getInstance("RSA/ECB/PKCS1Padding");
            rsa.init(Cipher.ENCRYPT_MODE, getPublicKey(publicKey));
            encryptedByte = rsa.doFinal(data.getBytes());
        }catch(Exception ex){
            ex.printStackTrace();
        }
        return Base64.getEncoder().encodeToString(encryptedByte);
    }

    public static String decrypt(String data, String privateKey) {
        Cipher rsa;
        byte[] decryptedByte = new byte[0];
        try {
            rsa = Cipher.getInstance("RSA/ECB/PKCS1Padding");
            rsa.init(Cipher.DECRYPT_MODE, getPrivateKey(privateKey));
            decryptedByte = rsa.doFinal(Base64.getDecoder().decode(data.getBytes()));
        }catch(Exception ex){
            ex.printStackTrace();
        }
        return new String(decryptedByte);
    }

    public static PublicKey getPublicKey(String base64PublicKey){
        PublicKey publicKey = null;
        try{
            X509EncodedKeySpec keySpec = new X509EncodedKeySpec(Base64.getDecoder().decode(base64PublicKey.getBytes()));
            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
            publicKey = keyFactory.generatePublic(keySpec);
        } catch (Exception e){
            e.printStackTrace();
        }
        return publicKey;
    }


    public static PrivateKey getPrivateKey(String base64PrivateKey){
        PrivateKey privateKey = null;
        try {
            PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(Base64.getDecoder().decode(base64PrivateKey.getBytes()));
            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
            privateKey = keyFactory.generatePrivate(keySpec);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return privateKey;
    }

    public static byte[] sign(String message, String privateKey) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] messageHash = md.digest(message.getBytes());
            Cipher cipher = Cipher.getInstance("RSA");
            cipher.init(Cipher.ENCRYPT_MODE, getPrivateKey(privateKey));
            return cipher.doFinal(messageHash);

        }catch ( Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static boolean verifySign(byte[]  encryptedMessageHash, String message, String publicKey) {
        try {
            Cipher cipher = Cipher.getInstance("RSA");
            cipher.init(Cipher.DECRYPT_MODE, getPublicKey(publicKey));
            byte[] decryptedMessageHash = cipher.doFinal(encryptedMessageHash);
            byte[] messageBytes = message.getBytes();
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] newMessageHash = md.digest(messageBytes);
            return Arrays.equals(decryptedMessageHash, newMessageHash);
        } catch ( Exception e) {
            e.printStackTrace();
        }
        return false;
    }
}

File: encryption-component/src/main/kotlin/infrastructure/EncryptUtils.kt
package infrastructure

import java.security.Key
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.spec.EncodedKeySpec
import java.security.spec.KeySpec
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.util.*
import javax.crypto.Cipher

object EncryptUtils {

    fun encrypt(message: String, publicKey: String): String {
        val encryptedByte: ByteArray? = cipherMessage(
                "RSA/ECB/PKCS1Padding",
            message.toByteArray(),
            Cipher.ENCRYPT_MODE,
            getPublicKey(publicKey)
        )
        return Base64.getEncoder().encodeToString(encryptedByte)
    }

    fun decrypt(message: String, privateKey: String): String {
        val decryptedByte: ByteArray? = cipherMessage(
                "RSA/ECB/PKCS1Padding",
            Base64.getDecoder().decode(message.toByteArray()),
            Cipher.DECRYPT_MODE,
            getPrivateKey(privateKey)
        )
        return String(decryptedByte!!)
    }

    fun getPublicKey(base64Key: String): Key {
        val keyFactory = KeyFactory.getInstance("RSA")
        val keySpec = X509EncodedKeySpec(Base64.getDecoder().decode(base64Key.toByteArray()))
        return getKey(keySpec,keyFactory::generatePublic)
    }

    fun getPrivateKey(base64Key: String): Key {
        val keyFactory = KeyFactory.getInstance("RSA")
        val keySpec = PKCS8EncodedKeySpec(Base64.getDecoder().decode(base64Key.toByteArray()))
        return getKey(keySpec, keyFactory::generatePrivate)
    }

    private fun getKey(keySpec: EncodedKeySpec, generateKey: (keySpec: KeySpec) -> Key): Key {
        return generateKey(keySpec)
    }

    fun signMessage(message: String, privateKey: String): ByteArray? {
        try {
            val hashedMessage = hashMessage(message)
            return cipherMessage("RSA", hashedMessage, Cipher.ENCRYPT_MODE, getPrivateKey(privateKey))
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    fun verifySign(encryptedMessageHash: ByteArray, message: String, publicKey: String): Boolean {
        try {
            val hashedMessage = hashMessage(message)
            val cipherHashedMessage = cipherMessage("RSA", encryptedMessageHash, Cipher.DECRYPT_MODE, getPublicKey(publicKey))
            return Arrays.equals(hashedMessage, cipherHashedMessage)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    private fun cipherMessage(transformation: String, message: ByteArray, cipherMode: Int, key: Key): ByteArray? {
        val rsa: Cipher
        var encryptedByte: ByteArray? = ByteArray(0)
        try {
            rsa = Cipher.getInstance(transformation)
            rsa.init(cipherMode, key)
            encryptedByte = rsa.doFinal(message)
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
        return encryptedByte
    }

    private fun hashMessage(message: String): ByteArray {
        val messageDigest = MessageDigest.getInstance("SHA-256")
        return messageDigest.digest(message.toByteArray())
    }

}